### PR TITLE
✨ Mark specific KCP machines with delete annotation for scaling down

### LINKED ISSUE #3910: Mark KCP machines for deletion
**User Story**
As a user I would like to mark a specific controlplane Machine to be deleted by a KCP after I scale it down.

**Detailed Description**
As an example, I have three controlplane Machines (a. b, c). Now I want to scale down KCP replicas to 1 but I want specifically  b and c Machines to be deleted after I run `kubectl scale kcp example --replicas=1`. As we discussed with @vincepri over the [slack](https://kubernetes.slack.com/archives/C8TSNPY4T/p1605129002289200), we can mark&delete(by scaling down) Machines by adding `cluster.x-k8s.io/delete-machine=`  annotation on them but only IF it is a worker Machine (controlled by MachineDeployment) but not controlplane Machine (controlled by KCP).

It would be nice to have similar way of marking specific controlplane Machines that we want to delete,  as currently it can be done for MachineDeployment Machines via `cluster.x-k8s.io/delete-machine=` annotation.

/kind feature

### PR DESCRIPTION
**Pull request by furkatgofurov7**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Adds a possibility to mark a specific control plane Machine to be deleted by a KCP when one or multiple machines are annotated by `delete-machine`  annotation.

**This PR:**
1. makes sure the oldest machine is deleted first when there is one or more machine is marked for deletion. It implements the prioritization logic in descending order with respect to rollout policies, while selecting machine for scale  down in the following order:
      - _outdatedMachines with the delete annotation_
      - _machines with the delete annotation_
      - _outdated machines_
      - _all machines_
2. adds a unit test.
3. updates the KCP proposal with a note describing the behavior for scale down with annotation.

Fixes #3910

### CODE REVIEW

File: controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go
Comment (vincepri) [inline controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go:33]: Could we just reuse the `DeleteMachineAnnotation` defined in the core types?

**Review by vincepri — CHANGES_REQUESTED** [review]

File: controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go
Comment (vincepri) [reply controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go:33]: I just realized that somehow that constant is defined in the `controllers` package. We should move it under the `api/v1alpha4` package instead, and then we should be able to reuse it here

Comment (fabriziopandini) [reply controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go:33]: +1

File: docs/proposals/20191017-kubeadm-based-control-plane.md
Comment (fabriziopandini) [inline docs/proposals/20191017-kubeadm-based-control-plane.md:340]: From what I understand, the annotation is a way to condition how to select which machine to delete when scaling down, but the annotation alone does not trigger scale down. 
If this is true, this sentence should be changed (or removed)

File: controlplane/kubeadm/internal/control_plane.go
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/control_plane.go:128]: Can we avoid to return an error value, given that it is always nil?

File: controlplane/kubeadm/controllers/scale.go
Comment (fabriziopandini) [inline controlplane/kubeadm/controllers/scale.go:141]:
```suggestion
	if withDeleteAnnotation, err := controlPlane.MachineWithDeleteAnnotation(machines); err == nil {
		return withDeleteAnnotation, nil
	}
```
Also given that error is always nil, should this check if there is at least one  machine withDeleteAnnotation?

File: controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go
Comment (furkatgofurov7) [reply controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go]: Thanks for the suggestion, I have made the changes accordingly.

**Comment by furkatgofurov7** [comment]
@vincepri @fabriziopandini thanks, I have addressed all your comments(I accidentally amended my commit, sorry for that). However, tests are failing, and was wondering if some of them are just a flake in CI or I am missing something?

**Comment by fabriziopandini** [comment]
@furkatgofurov7 
Failure in pull-cluster-api-apidiff-main is ok (it just report you moved the annotation).
Some test are not flakes
- lint errors in pull-cluster-api-make-main 
- following errors in pull-cluster-api-test-main 
```
--- FAIL: TestSelectMachineForScaleDown (0.00s)
    --- FAIL: TestSelectMachineForScaleDown/when_there_is_a_machine(s)_marked_with_incorrect_delete_annotation_key,_it_returns_only_machine(s)_with_proper_annotation (0.00s)
```

File: controllers/machineset_delete_policy.go
Comment (vincepri) [inline controllers/machineset_delete_policy.go:36]: Revert this, after #3955 merges, we should rebase and this deprecated annotation is going to be removed from the codebase

File: controlplane/kubeadm/controllers/scale.go
Comment (vincepri) [inline controlplane/kubeadm/controllers/scale.go:140]: Let's move this at the top of this function, on line 134 given that the logic here effectively ignores any other criteria to pick a Machine to scale down, and always prefers the Machine that the user selected.

File: controlplane/kubeadm/internal/control_plane.go
Comment (vincepri) [inline controlplane/kubeadm/internal/control_plane.go:133]:
```suggestion
```

File: controlplane/kubeadm/internal/control_plane.go
Comment (vincepri) [inline controlplane/kubeadm/internal/control_plane.go:132]: Hm, could we instead return the whole list of annotated machines here? And let the logic above pick the one in the failure domain with most machines?

Comment (furkatgofurov7) [reply controlplane/kubeadm/internal/control_plane.go:132]: @vincepri we were discussing this in the issue itself and kind of agreed to return the oldest machine first. Are there any objections/drawbacks of don't it this way? would you mind elaborating more if so?  Thanks

Comment (vincepri) [reply controlplane/kubeadm/internal/control_plane.go:132]: `MachineInFailureDomainWithMostMachines` (what a mouthful 😂), picks the Machine that's in the failure domain (e.g. availability zone on AWS) with the highest count of machines in it, and then picks the oldest one from that list

File: controlplane/kubeadm/controllers/scale.go
Comment (vincepri) [reply controlplane/kubeadm/controllers/scale.go:137]: If we remove the .Oldest call inside `MachineWithDeleteAnnotation`, we can return the whole list from that function and overwrite the `machines` slice here. We should also move back this block after the `outdatedMachines` one, so it overwrites the list.

Then we can return normally

**Comment by furkatgofurov7** [comment]
> @furkatgofurov7
> Failure in pull-cluster-api-apidiff-main is ok (it just report you moved the annotation).
> Some test are not flakes
> 
>     * lint errors in pull-cluster-api-make-main
> 
>     * following errors in pull-cluster-api-test-main
> 
> 
> ```
> --- FAIL: TestSelectMachineForScaleDown (0.00s)
>     --- FAIL: TestSelectMachineForScaleDown/when_there_is_a_machine(s)_marked_with_incorrect_delete_annotation_key,_it_returns_only_machine(s)_with_proper_annotation (0.00s)
> ```

@fabriziopandini @vincepri I have addressed all your comments, thanks for the review. Can you please suggest how can we fix pull-cluster-api-apidiff-main test?

**Comment by fabriziopandini** [comment]
Failure in pull-cluster-api-apidiff-main is ok (it just report you moved the annotation); the job is not required, so it will not block this PR.

File: controlplane/kubeadm/controllers/scale.go
Comment (detiber) [inline controlplane/kubeadm/controllers/scale.go:244]: Is the intention here to prioritize outdated Machines over Machines marked for deletion by annotation?

This would mean effectively that the annotation would be ignored if applied to a newer machine during the process of a rollout of changes until that rollout is complete.

I'm wondering if this will cause confusion for users.

I think it might make more sense to prioritize:
- any outdatedMachines with the delete annotation
- any machines with the delete annotation
- outdated machines
- all machines

```suggestion
	switch {
	case controlPlane.MachineWithDeleteAnnotation(outdatedMachines).Len() > 0:
		machines = controlPlane.MachineWithDeleteAnnotation(outdatedMachines)
	case controlPlane.MachineWithDeleteAnnotation(machines).Len() > 0:
		machines = controlPlane.MachineWithDeleteAnnotation(machines)
	case outdatedMachines.Len() > 0:
		machines = outdatedMachines
	}
```

File: docs/proposals/20191017-kubeadm-based-control-plane.md
Comment (detiber) [inline docs/proposals/20191017-kubeadm-based-control-plane.md:344]: This covers the simplistic case related to scale down, depending on the prioritization question I posed above it would also be good to include information how a rollout would be affected by the presence of the annotation.

File: controlplane/kubeadm/controllers/scale_test.go
Comment (detiber) [inline controlplane/kubeadm/controllers/scale_test.go:312]: It would be good to also add test cases covering the expected behavior when an annotated machine is not in outdatedMachines but is part of annotatedControlPlane

**Comment by detiber** [comment]
/hold
holding pending clarification around prioritization of the deletion annotation wrt rollout of changes.

File: controlplane/kubeadm/controllers/scale.go
Comment (vincepri) [reply controlplane/kubeadm/controllers/scale.go:244]: That should be what the current code does, given that we replace the splice in each `if` block, but your switch case solution looks more readable to me 👍

Comment (detiber) [reply controlplane/kubeadm/controllers/scale.go:244]: @vincepri there is a slightly different effect if the delete annotation is added to a machine that is up to date according to our rollout policies while there are any outdated machines that still exist.

The prioritization of the current code in the PR is:
- any outdatedMachines with the delete annotation
- outdated machines
- any machines with the delete annotation
- all machines

compared to my suggestion:
- any outdatedMachines with the delete annotation
- any machines with the delete annotation
- outdated machines
- all machines

Comment (vincepri) [reply controlplane/kubeadm/controllers/scale.go:244]: Oh I see what you mean, yes that makes sense! ✅

Comment (furkatgofurov7) [reply controlplane/kubeadm/controllers/scale.go]: That makes sense to me as well, thanks for suggestion @detiber. I have addressed your comments. The original idea was giving priority to any machines(outdatedMachines or not) with annotation a top priority, as you have clarified above.  WDYT of this version? PTAL.

**Comment by detiber** [comment]
/lgtm
/hold cancel

@furkatgofurov7 thank you for the fixes, these changes lgtm

**Comment by furkatgofurov7** [comment]
Thanks a lot, @detiber @fabriziopandini @vincepri for reviews/suggestions!

**Pull request merged**
